### PR TITLE
adding first shot at k8s configs for matomo (logging).

### PR DESCRIPTION
**Pull request by vsoch**
This is my first time writing these configs, so apologies in advance for things I don't know, and no worries about giving feedback to make them correct/better. I added some questions as comments, and information about the resources we need is at:

https://matomo.org/docs/requirements/#recommended-servers-sizing-cpu-ram-disks

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

File: k8s/logs-spack-io/deployments.yaml [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: matomo-config
+  labels:
+    app: matomo
+    environment: matomo
+data:
+  nginx.conf: |
```
Comment (tgamblin) [inline k8s/logs-spack-io/deployments.yaml:193]: Is it possible to put files like this one in the same directory rather than inlining them in the YAML?

Comment (vsoch) [reply k8s/logs-spack-io/deployments.yaml:193]: I'm not sure how k8s works but my basic understanding is that you'd have to create some kind of volumeMount to be able to do that. I was glancing at different examples and saw this method as being pretty common. But I defer to the expertise of @opadron !

Comment (tgamblin) [reply k8s/logs-spack-io/deployments.yaml:193]: I just mean in this repository -- I'm assuming that at the time you submit these files to k8s that any files in this repo are available and can be uploaded along with the YAML.  @opadron?

File: k8s/logs-spack-io/certificates.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: tls-www
```
Comment (opadron) [inline k8s/logs-spack-io/certificates.yaml:5]: need to add the namespace here

Comment (opadron) [reply k8s/logs-spack-io/certificates.yaml:5]: Also, can we switch the name to something like `tls-logs`, to more easily distinguish it from the certs used for spack.io?

File: k8s/logs-spack-io/deployments.yaml [hunk 3]
```diff
@@ -0,0 +1,352 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: logs-spack-io-redirect
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:5]: I'm assuming you pulled the `-redirect` part of this name from the spack.io config?  If so, I would skip all of the redirect stuff, unless this service is actually redirecting traffic somewhere?

Comment (opadron) [reply k8s/logs-spack-io/deployments.yaml:5]: So, `logs-spack-io` would be a better name.

File: k8s/logs-spack-io/deployments.yaml [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+  namespace: logs-spack-io
+  labels:
+    app: logs-spack-io
+    svc: web-redirect
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:9]: Probably want to change this to simply `web`.  I've been using `svc: web` for components that directly handle web traffic.

File: k8s/logs-spack-io/deployments.yaml [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+    environment: matomo
+
+spec:
+  selector:
+    matchLabels:
+      app: logs-spack-io
+      svc: web-redirect
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:16]: `svc: web`

File: k8s/logs-spack-io/deployments.yaml [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+      environment: matomo
+  replicas: 2
+  template:
+    metadata:
+      labels:
+        app: logs-spack-io
+        svc: web-redirect
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:23]: `svc: web`

File: k8s/logs-spack-io/deployments.yaml [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        - name: AWS_ACCESS_KEY_ID
+          valueFrom:
+            secretKeyRef:
+              name: spack-aws-credentials
+              key: aws_access_key_id
+        - name: AWS_SECRET_ACCESS_KEY
+          valueFrom:
+            secretKeyRef:
+              name: spack-aws-credentials
+              key: aws_secret_access_key
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:60]: For AWS operations, we've switched to using kubernetes service accounts with an attached IAM Role.  I'll sync with you out-of-band to work this part out.

File: k8s/logs-spack-io/deployments.yaml [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+        - name: MYSQL_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: logs-spack-io-secrets
+              key: mysql_password
+        ports:
+        - containerPort: 3306
+          name: mysql
+
+        # Important - this is recommended to be 500GB SSD - can we do that?
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:136]: Yes, though we should get together to work out some of the details.  I want to make sure you get the best storage available.

File: k8s/logs-spack-io/deployments.yaml [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+            memory: "3Gi"
+          requests:
+            cpu: "0.25"
+            memory: "250Mi"
+      volumes:
+      - name: config
+        configMap:
+          name: matomo-config
+
+      # TODO how do we update this for a filesystem?
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:169]: This part will need to be replaced with a persistent volume.

File: k8s/logs-spack-io/deployments.yaml (hunk 1 above)
Comment (opadron) [reply k8s/logs-spack-io/deployments.yaml:193]: @tgamblin unfortunately not.  You can pass local files as cli arguments to kubectl when creating configmaps, but `flux` only works directly with the raw manifests, so inlining is the only way.

File: k8s/logs-spack-io/deployments.yaml [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+          user: admin
+          secretRef:
+            name: logs-spack-io-secrets
+            key: logs-spack-io-secrets
+          readOnly: false
+      nodeSelector:
+        "beta.kubernetes.io/instance-type": "t2.2xlarge"
+
+
+---
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:184]: Let's split off each of the distinct k8s objects to their own files. (`k8s/logs-spack-io/configmaps.yaml`)

File: k8s/logs-spack-io/deployments.yaml [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+          fastcgi_pass backend;
+        }
+      }
+    }
+  php.ini: |
+    memory_limit=1536M
+    max_execution_time=1440
+
+
+---
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:285]: `k8s/logs-spack-io/cronjobs.yaml`

File: k8s/logs-spack-io/deployments.yaml [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+          - name: html-files
+            awsElasticBlockStore:
+              # Can this be from secrets too?
+              volumeID: "volume-xxx"
+              fsType: ext4         
+          - name: migration-logs
+            awsElasticBlockStore:
+              # Can this be from secrets too?
+              volumeID: "volume-xxx"
+              fsType: ext4
```
Comment (opadron) [inline k8s/logs-spack-io/deployments.yaml:352]: We'll work out the storage side of this equation.

File: k8s/logs-spack-io/secrets.yaml [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  annotations:
+      fluxcd.io/ignore: "true"
+data:
+  # These are all fake secrets!
+  matomo_database_tables_prefix: hakunamatata
+  mysql_root_password: hakunamatata
+  mysql_host: hakunamatata
+  mysql_user: hakunamatata
+  mysql_password: hakunamatata
+  mysql_database: hakunamatata
```
Comment (opadron) [inline k8s/logs-spack-io/secrets.yaml:16]: What a wonderful phrase! 🎉

File: k8s/logs-spack-io/services.yaml [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+##     svc: web-redirect
+## spec:
+##   type: ClusterIP
+##   ports:
+##   - name: http
+##     port: 80
+##     targetPort: http
+##   selector:
+##     app: spack-io
+##     svc: web-redirect
```
Comment (opadron) [inline k8s/logs-spack-io/services.yaml:37]: This can all go.

**Review by opadron — CHANGES_REQUESTED** [review]
Good start!  I'll get in touch over Slack to help you

- [ ] get set up with a service account so you won't have to pass AWS credentials directly, and
- [ ] get some storage that should meet your application's needs

File: k8s/logs-spack-io/certificates.yaml (hunk 2 above)
Comment (vsoch) [reply k8s/logs-spack-io/certificates.yaml:5]: yep!

**Comment by vsoch** [comment]
@opadron updates!

https://github.com/spack/spack-infrastructure/pull/75/commits/67e6890dd8fe5c962a136f513563bd08dd65ab6f

I'm in a workshop for another few hours but will be around to talk storage on slack after that! Is today a good day to ping you, or would you like another day?

**Comment by opadron** [comment]
@vsoch I'll ping you on slack!

File: k8s/logs-spack-io/deployments.yaml (hunk 1 above)
Comment (tgamblin) [reply k8s/logs-spack-io/deployments.yaml]: ok sounds good to me in that case!

**Comment by opadron** [comment]
We're gonna shelf this for now since we're looking into using an AWS-managed solution for analytics.

**Pull request closed without merging**